core/middlewares/with-auth.ts

```diff
@@ -1,17 +1,68 @@
-import { auth, signIn } from '~/auth';
+import { NextResponse } from 'next/server';
+
+import { auth, signIn, signOut } from '~/auth';
+import { client } from '~/client';
+import { graphql } from '~/client/graphql';
 
 import { type MiddlewareFactory } from './compose-middlewares';
 
+// Path matcher for any routes that require authentication
+const protectedPathRegExp = /^\/(account).*$/;
+
+async function checkTokenValidity(customerAccessToken: string, email?: string): Promise<boolean> {
+  const result = await client.fetch({
+    document: graphql(`
+      query ValidateToken {
+        customer {
+          email
+        }
+      }
+    `),
+    customerAccessToken,
+    fetchOptions: { cache: 'no-store' },
+  });
+
+  return result.data.customer?.email === email;
+}
+
+function redirectToLogin(url: string) {
+  return NextResponse.redirect(new URL('/login', url), { status: 302 });
+}
+
 export const withAuth: MiddlewareFactory = (next) => {
   return async (request, event) => {
     // @ts-expect-error: The `auth` function doesn't have the correct type to support it as a MiddlewareFactory.
     const authWithCallback = auth(async (req) => {
+      const isProtectedRoute = protectedPathRegExp.test(req.nextUrl.pathname);
+      const isGetRequest = req.method === 'GET';
+
       if (!req.auth) {
         await signIn('anonymous', { redirect: false });
 
+        if (isProtectedRoute && isGetRequest) {
+          return redirectToLogin(req.url);
+        }
+
         return next(req, event);
       }
 
+      const { email, customerAccessToken } = req.auth.user ?? {};
+
+      if (customerAccessToken && isProtectedRoute) {
+        // To avoid unnecessary expense, only check token validity on routes that require authentication
```

Comment:
Technically this doesn't just apply to account related routes but also products and categories. Since we don't want to add additional overhead fetches, and we do some fetching in the `with-routes` middleware, I am tempted that we try to handle an invalid token there 🤔

Reply:
After chatting, we should be doing this check at the DAL and middleware should be specific to protecting the routes if it's authed or not.

Reply:
@chanceaclark For sure. What concerns me there is that it's possible for the `customerAccessToken` to exist, even though it's invalid. This would allow access to protected routes and you would just end up at a 404 page again like before.

In order to avoid that, we would still need the token validation here so that we can redirect to the login page if your current token isn't valid 🤔 Unless we are able to remove the customer access token at the DAL before the request reaches middleware, but I'm not sure if that would be possible.

Reply:
I think the DAL should be the one to handle if a user has the ability to access data, if not log them out to clear the session. Middleware should be the one to basically check if the session exists or not and allow users to navigate there or not. Like mentioned in standup, I think our Storefront GQL should return an error (with partial data) if using an invalid customer access token with the request. I think it's a good improvement to consumers of our API too here.